### PR TITLE
Rename ring buffer to com.canonical.lxd 

### LINKED ISSUE #12149: Rename ring buffer to com.canonical.lxd
# Issue description

Currently, the ring buffer is still named `org.linuxcontainers.lxd`. This should be renamed to `com.canonical.lxd`. Renaming the ring buffer will break existing VMs as the path is hard-coded in the lxd-agent setup scripts.

### PR DESCRIPTION
**Pull request by MusicDin**
This PR updates the name of ring buffer from `org.linuxcontainers.lxd` to `com.canonical.lxd`.

With this change in place, restart of existing VMs will be required. Otherwise, (old) lxd-agent will write data to the `org.linuxcontainers.lxd` qemu serial device that is not connected to the new ring buffer from which lxd server will try to read. After VM is restarted, new lxd-agent will be copied from the shared host filesystem to the VM.

Fixes #12149

### CODE REVIEW

**Comment by mihalicyn** [comment]
Looks good to me!

And also thanks for adding a great explanatory comments.

**Review by mihalicyn — DISMISSED** [review]

**Comment by mihalicyn** [comment]
We still need to test this very-very carefully to be 100% that nothing was missed.

File: lxd/instance/drivers/driver_qemu.go [hunk 1]
```diff
@@ -2556,13 +2558,19 @@ chown -R root:root "${PREFIX}"
...
 
-	// Udev rules
 	err = os.MkdirAll(filepath.Join(configDrivePath, "udev"), 0500)
 	if err != nil {
 		return err
 	}
 
-	lxdAgentRules := `ACTION=="add", SYMLINK=="virtio-ports/org.linuxcontainers.lxd", TAG+="systemd", ACTION=="add", RUN+="/bin/systemctl start lxd-agent.service"`
+	// Udev rules that trigger the start of the lxd-agent.service when qemu serail devices (symlinks
+	// in virtio-ports) appear.
```
Comment (simondeziel) [inline lxd/instance/drivers/driver_qemu.go:2567]:
```suggestion
	// udev rules to start the lxd-agent.service when QEMU serial devices (symlinks in virtio-ports) appear.
```

Comment (simondeziel) [reply lxd/instance/drivers/driver_qemu.go:2567]: Actually, would it make sense to have the comment in the udev rule file itself?

File: lxd/instance/drivers/driver_qemu_templates.go [hunk 2]
```diff
@@ -177,19 +177,33 @@ func qemuSerial(opts *qemuSerialOpts) []cfgSection {
 		comment: "Virtual serial bus",
 		entries: qemuDeviceEntries(&entriesOpts),
 	}, {
+		// Ring buffer used by an lxd agent to write it's status in it. LXD server will read
```
Comment (simondeziel) [inline lxd/instance/drivers/driver_qemu_templates.go:180]:
```suggestion
		// Ring buffer used by the lxd agent to report (write) its status to. LXD server will read
```

**Review by simondeziel — APPROVED** [review]
LGTM with some tiny nitpicks.

File: lxd/instance/drivers/driver_qemu.go (hunk 1 above)
Comment (MusicDin) [reply lxd/instance/drivers/driver_qemu.go:2567]: Yes, I'll add them into configuration within lxd-agent-loader package.

**Comment by tomponline** [comment]
@simondeziel @MusicDin would we be able to add some tests to lxd-ci that would check that the agent is working on an older image (e.g. jammy assuming the SRU won't go through) is working with the newer LXD?

And also that the agent is working in an image for a newer version of Ubuntu (once our changes land in lxd-agent-loader package) on an older (5.0/stable) LXD?

**Review by tomponline — DISMISSED** [review]
LGTM thanks!

What was the outcome of the test to statefully restore a VM running an older lxd-agent that talks to the older serial device?

**Comment by MusicDin** [comment]
Steps that I did (Also see the spec where I tried to document this https://discourse.ubuntu.com/t/rename-lxd-ring-buffer-to-com-canonical-lxd/40499):
```
$ lxc launch ubuntu:22.04 v1 --vm
$ lxc snapshot v1 mysnap --stateful

# >> Upgrade LXD

# This will work, because VM state is not recovered.
$ lxc restore v1 mysnap
$ lxc start v1

# This will NOT work, because old LXD agent is restored as well and it is not able
# to communicate with the LXD server through a common ring buffer.
$ lxc restore v1 mysnap --stateful
Error: Failed restoring state from "/var/snap/lxd/common/lxd/virtual-machines/v1/state": Monitor is disconnected
```

**Comment by tomponline** [comment]
> Steps that I did (Also see the spec where I tried to document this https://discourse.ubuntu.com/t/rename-lxd-ring-buffer-to-com-canonical-lxd/40499):
> 
> ```shell
> $ lxc launch ubuntu:22.04 v1 --vm
> $ lxc snapshot v1 mysnap --stateful
> 
> # >> Upgrade LXD
> 
> # This will work, because VM state is not recovered.
> $ lxc restore v1 mysnap
> $ lxc start v1
> 
> # This will NOT work, because old LXD agent is restored as well and it is not able
> # to communicate with the LXD server through a common ring buffer.
> $ lxc restore v1 mysnap --stateful
> Error: Failed restoring state from "/var/snap/lxd/common/lxd/virtual-machines/v1/state": Monitor is disconnected
> ```

Hrm, are you sure the failure to restore the snapshot was due to the ring buffer? Did you get an error from the qemu logs?

**Comment by MusicDin** [comment]
Good point, need to verify that

File: lxd/instance/drivers/driver_qemu.go [hunk 3]
```diff
@@ -2500,16 +2502,16 @@ Restart=on-failure
 RestartSec=5s
 StartLimitInterval=60
 StartLimitBurst=10
-
-[Install]
-WantedBy=multi-user.target
```
Comment (simondeziel) [inline lxd/instance/drivers/driver_qemu.go:2505]: With that `[Install]` gone we rely on `udev` as the only trigger. As such, I think we should update the `install.sh` script to also clean the old symlink:

```
# ll /etc/systemd/system/multi-user.target.wants/lxd-agent.service 
lrwxrwxrwx 1 root root 37 Oct 27 02:35 /etc/systemd/system/multi-user.target.wants/lxd-agent.service -> /lib/systemd/system/lxd-agent.service
```

Comment (MusicDin) [reply lxd/instance/drivers/driver_qemu.go:2505]: Good point, do you mean `lxd-agent-setup` script?

**Comment by MusicDin** [comment]
> Hrm, are you sure the failure to restore the snapshot was due to the ring buffer? Did you get an error from the qemu logs?

You were right Tom, my assumption was wrong. According to @mihalicyn, the issue is related to non-existing state of the new serial device.

Here is the instance log:
```
qemu-system-x86_64: Failed to load virtio-console:virtio
qemu-system-x86_64: error while loading state for instance 0x0 of device '0000:00:01.0:00.5/virtio-console'
qemu-system-x86_64: load of migration failed: Invalid argument
```

File: lxd/instance/drivers/driver_qemu.go (hunk 3 above)
Comment (simondeziel) [reply lxd/instance/drivers/driver_qemu.go:2505]: No, you can search for `lxdConfigShareInstall` and you'll see which file I'm referring to :)

Comment (MusicDin) [reply lxd/instance/drivers/driver_qemu.go:2505]: Ahh yeah, thanks. Will add it.

**Comment by tomponline** [comment]
OK thats fine then as we wont need to consider restoring an actively running older agent.

**Review by tomponline — APPROVED** [review]

**Pull request merged**